### ISSUE TITLE
FR: Offer compressed alternatives for "small Tailscale" binaries

### ISSUE THREAD

**Issue by iGerman00** [tailscale/tailscale#8613]
### What are you trying to do?

I have an OpenWRT router with ~100MB of flash storage, using up 30MB of that is the tailscale binary. Its architecture is an unusual `mips-siflower`, however the `mipsle` (mipsel) build works on it just fine.

Compressing it myself with UPX 4.0.2 didn't lead to much of anything good and it refused to launch "two instances" of the combined binary, dropping a segfault, but uncompressed works fine. The daemon was running, but the CLI refused to, so clearly something has gone wrong

### How should we solve this?

- Include a download for pre-compressed "as small as possible" binaries
- Verify if it works in a configuration such as a running daemon and the CLI simultaneously before shipping it
- Offer two small, compressed, but separate binaries, if that is UPX' fault (alternative)

### What is the impact of not solving this?

- No functionality on smaller routers, embedded machines and similar devices where persistent storage is lacking
- Potential for user error and poor experience during compression and/or building an own version of the tailscale binary from source

### Anything else?

https://gist.github.com/iGerman00/b45f78685f7db0fa264edf7a0d81bd50 - Guide by me on how to install tailscale on such a router, which produced the UPX instancing bug

https://github.com/lanrat/openwrt-tailscale-repo - Hosted tailscale repository for OpenWRT

**Comment by DentonGentry** [comment]
> I have an OpenWRT router with ~100MB of flash storage, using up 30MB of that is the tailscale binary.

It should not be 30 MBytes, it should be more like 4 MBytes. You're using https://tailscale.com/kb/1207/small-tailscale/ ?

**Comment by iGerman00** [comment]
> > I have an OpenWRT router with ~100MB of flash storage, using up 30MB of that is the tailscale binary.
> 
> It should not be 30 MBytes, it should be more like 4 MBytes. You're using https://tailscale.com/kb/1207/small-tailscale/ ?

Yes. You can check my guide to see the exact build process I used. You can also check the repository that I linked to see how that’s built

**Comment by DentonGentry** [comment]
The problem is that the compressed binary segfaults?

So the feature request is to first fix compressed binaries for these platforms, and then supply pre-built binaries?

**Comment by iGerman00** [comment]
That would be ideal, but I’m not sure where the problem with segfaulting is - whether it’s something from Tailscale, UPX, Go or this particular platform, hence this issue to try and figure that out. I might open another issue on UPX’s repo later. 

At the very least a pre-compressed binary is nice to have, with the assumption that it “looks good” on your side, then we can discuss downstream with other parties involved

**Comment by hazarjast** [comment]
It seems there are some known MIPS issues still for UPX, maybe related?
https://github.com/upx/upx/issues/788
